### PR TITLE
Upgrade to Docusaurus v3

### LINKED ISSUE #347: Update to React 18
We currently use React 17 due to the Docusaurus limitation:

- https://github.com/stylelint/stylelint.io/blob/ca1ce93e75069278a97a62780d089f8cf23695b4/.github/dependabot.yml#L11-L16
- https://github.com/facebook/docusaurus/issues/7264

But Docusaurus 3 will likely update to React 18. See https://github.com/facebook/docusaurus/milestone/20?closed=1

We need to wait for the new version of Docusaurus.

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Closes #347

> Is there anything in the PR that needs further explanation?

See the upgrade guide: https://docusaurus.io/docs/migration/v3

Summary:
- Bump dependencies: Docusaurus and React
- Move `@docusaurus/*` from `devDependencies` to `dependencies` because the official doc describes so.
- Rewrite `docusaurus.config.js` in ESM

### CODE REVIEW

File: scripts/generate-docs.mjs [hunk 1]
```diff
@@ -82,8 +82,8 @@ function wrapProblemExample() {
  */
 function convertToAdmonitions() {
 	const gfmToAdmonitions = new Map([
-		['Note', 'note'],
-		['Warning', 'caution'],
+		['Note', 'info'],
+		['Warning', 'warning'],
```
Comment (ybiquitous) [inline scripts/generate-docs.mjs:86]: [note] I feel `info` is nearer to GFM than `note`. See below:
- https://docusaurus.io/docs/markdown-features/admonitions
- https://github.com/orgs/community/discussions/16925

---

[note] `caution` is deprecated in v3. See below:
- https://docusaurus.io/docs/migration/v3#admonition-warning

File: src/pages/demo.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import React, { useEffect, useRef } from 'react';
+import { useEffect, useRef } from 'react';
```
Comment (ybiquitous) [inline src/pages/demo.js:1]: [note] Docusaurus v3 now uses the React 18 "automatic" JSX runtime.
See https://docusaurus.io/docs/migration/v3#automatic-jsx-runtime

File: package.json [hunk 3]
```diff
@@ -52,7 +52,8 @@
     },
     "extends": [
       "stylelint",
-      "plugin:react/recommended"
+      "plugin:react/recommended",
+      "plugin:react/jsx-runtime"
```
Comment (ybiquitous) [inline package.json:56]: [note] `plugin:react/jsx-runtime` is for the new JSX transform.
See https://github.com/jsx-eslint/eslint-plugin-react#configuration-legacy-eslintrc-

File: package.json [hunk 4]
```diff
@@ -16,14 +16,14 @@
...
-    "gendoc": "node ./scripts/generate-docs.mjs docs",
+    "gendoc": "node ./scripts/generate-docs.mjs docs/",
     "lint": "npm-run-all --parallel lint:*",
     "lint:css": "stylelint \"**/*.css\" --ignore-path .gitignore ",
     "lint:formatting": "prettier . --check",
     "lint:js": "eslint . --ignore-path .gitignore  --ext .js,.mjs",
     "lint:md": "remark . --quiet --frail --ignore-path .gitignore",
     "prepare": "husky install",
-    "start": "npm run gendoc && docusaurus start --no-open",
+    "start": "npm run gendoc && docusaurus start",
```
Comment (ybiquitous) [inline package.json:26]: [note] `--no-open` is no longer needed because the same tab is opened.

File: scripts/generate-docs.mjs [hunk 5]
```diff
@@ -215,6 +215,15 @@ function main(outputDir) {
...
 
+		// TODO: Prevent the parse error. Remove this code when the error is fixed.
+		if (outputFile.endsWith('CHANGELOG.md')) {
+			output = output.replace(' #{&}', ' `#{&}`');
+
+			console.warn(
+				`Warning: The parse error in '${outputFile}' was prevented: 'Could not parse expression with acorn'.`,
+			); // eslint-disable-line no-console
+		}
+
```
Comment (ybiquitous) [inline scripts/generate-docs.mjs:226]: [note] I'll fix Stylelint's `CHANGELOG.md`.
See https://github.com/stylelint/stylelint/blob/c123cf1901db534a0186d107a8e9e650a0a52fa5/CHANGELOG.md?plain=1#L914

**Review by jeddy3 — COMMENTED** [review]
I saw that Docuasurus v3 was released when catching up on my RSS feeds, and it's terrific to see you've already opened a PR @ybiquitous!

It's looking great. 

On the [deploy preview](https://deploy-preview-357--stylelint.netlify.app/user-guide/rules), it seems some of our custom CSS isn't active:

Screenshot 2023-11-05 at 11 13 18

The following aren't hidden:
- the 2nd (auto-generated) "Rules" navigation item
- the empty table headers on the Rules page

However, they are hidden when I run the site locally, which is strange. Any ideas?

---

Should we also adjust the `"eslintConfig": { "settings": { "react": { "version": "17.02" }}}` version too?

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for the review!

> However, they are hidden when I run the site locally, which is strange. Any ideas?

Um..., the preview seems to be working in my browser. 🤔 

image

---

> Should we also adjust the `"eslintConfig": { "settings": { "react": { "version": "17.02" }}}` version too?

Good catch. Fixed with 345edf5.

**Review by jeddy3 — APPROVED** [review]
> Um..., the preview seems to be working in my browser. 🤔

Very odd. It's probably something on my end then.

Marking as approved.

**Pull request merged**